### ISSUE TITLE
Request: myriahedral map projection

### ISSUE THREAD

**Issue by espinielli** [d3/d3-geo-projection#67]
Given I have seen issues with requests for "fancy" map projections, why not proposing the fanciest?

[Myriahedral map projections](https://www.win.tue.nl/~vanwijk/myriahedral/) by Prof. [Jarke van Wijk](https://www.win.tue.nl/~vanwijk/)

![image](https://cloud.githubusercontent.com/assets/891692/16610827/1b19ae0c-435e-11e6-8855-cc8b4928f129.png)

**Comment by Fil** [comment]
<img width="619" alt="capture d ecran 2017-08-07 a 15 21 25" src="https://user-images.githubusercontent.com/7001/29028333-383693a6-7b84-11e7-8db4-a076205c1a37.png">

I'm confident we'll get there — sooner or later :)

**Comment by Fil** [comment]
<img width="674" alt="myriahedral projection" src="https://user-images.githubusercontent.com/7001/29041479-451ceb22-7bb3-11e7-9b34-a6960e2173db.png">

**Comment by Fil** [comment]
<img width="719" alt="icosahedral projection" src="https://user-images.githubusercontent.com/7001/29045404-c787f9b6-7bc3-11e7-8542-09b56e96850c.png">

icosahedral (#87) is sort-of a consequence of this

**Comment by Fil** [comment]
I had not read prof. Jarke van Wijk (JvW)'s article in details, and mainly used the images as a source of inspiration. Now I'm trying to figure out how far my implementation is from his.

Being lazy, I didn't want to create an algorithm for the meshes coordinates, and used geoVoronoi instead. So, I create the mesh of cells _from their centers_ (JvW constructs the "dual mesh" from the initial mesh). Then, from the geo Delaunay list of links I extract a spanning tree with a minimal total length (this could be any other weight function, to reflect geographic features). For this, I used Kruskal's algorithm not Prim's (not that it matters).

The [map above](https://github.com/d3/d3-geo-projection/issues/67#issuecomment-320750994) is defined by the centroids of each feature (countries) from a geojson file — and as there are more countries on the continents than in the oceans, it tends to use the continents as "valleys" of the minimal spanning tree.

Having geoVoronoi cells, I center the gnomonic projection for each face at its corresponding voronoi center. This has the nice property that the centers of two linked faces are equidistant from their common edge, so the scale must be the same on the two faces. As a consequence, it is constant throughout the mesh. (This also help explain why the paper model folds back perfectly, at least in one case.)

<img width="1361" alt="voronoi earth paper toy" src="https://user-images.githubusercontent.com/7001/29129692-ec17c6ac-7d27-11e7-876d-98d64971b289.png">

It does not mean that the projection is equivalent. However, the scale is the same locally around each of the initial centers. And the distortion is capped by the maximum distance between a cell's perimeter and its center (which MUST be  < 90° for the gnomonic) — which converges to zero as the density of cells grows.

Here's another map where I added one step of "Lloyd's relaxation" to grow the smaller cells and avoid having too much distorsion. I also removed the smaller countries from the mesh, so there were less cuts.

<img width="819" alt="voronoi projection" src="https://user-images.githubusercontent.com/7001/29129724-0319c134-7d28-11e7-8e36-37dc2ba72210.png">


An advantage of using geoVoronoi is that I don't have to create smaller triangles to accomodate polygonal cells as JvW does ("by inserting interior edges with very high weights", p. 34). As long as the maximum distance between a cell's perimeter and center is < 90°, the projection will map each polygon without further splicing, and this is why it's feasible to draw each polygon with a specific color.

There are things that I have yet to do (and may be possible or not):

- In particular, find a way to recreate figures similar to those defined by JvW (I particularly like "the coastline of the Earth" — figure 12). This means experimenting with recipes to define the centers and the weights of the links (or using something else than the MST to define the relevant spanning tree).

- Also, d3.geoPolyhedral doesn't (yet) allow "partial unfolding", so we can't make this very nice movie of the map slowly going from orthographic to unfolded. (Projection transition? three.js?) There's lots of experimentation to do here too.

To sum up, I don't think that my system is implementing myriahedral projections in the way they are defined in JvW's article, and while some results look very similar, the approach is quite different.

**Maybe using a different name would be more appropriate? (polyhedralVoronoi?)**

PS: For fun I generated a mesh from a [Fibonacci quasi-random distribution of points on the sphere](https://bl.ocks.org/Fil/955da86d6a935b26d3599ca5e344fb38), and called the result "fibonacci projection". It makes no sense but it's a great test case for https://github.com/d3/d3-geo/pull/108
<img width="661" alt="fibonacci sphere projection" src="https://user-images.githubusercontent.com/7001/29130015-e5b36356-7d28-11e7-8e02-da8d0dc382b6.png">

**Comment by jrus** [comment]
Doesn’t seem to be quite as fancy as the versions in the paper, but very nice work!

**Comment by Fil** [comment]
Fancier is on the way :)

Using a spanning tree focused on the oceans:

<img width="675" alt="voronoi coastline" src="https://user-images.githubusercontent.com/7001/29399127-46309c80-8328-11e7-9e50-5c37f2122161.png">

**Comment by Fil** [comment]
& with a focus on land masses 
<img width="765" alt="voronoi continents" src="https://user-images.githubusercontent.com/7001/29401124-3c677c34-8330-11e7-9b02-1d45b165e1dc.png">

**Comment by Schafknuddler** [comment]
I don't know how you managed to create your Myriahedral projections, but I am interested in creating such a projection myself and haven't found any tools that can do it, even though it would be easy to use, since you only really need to input a weighted mercator-projected map and the amount of subdivisions, and the computer spits out the optimal Myriahedral projection. If there is any framework like that already out there, please let me know. I am not desperate enough to do it myself.. yet...

**Comment by severo** [comment]
I guess it's this one: https://observablehq.com/@fil/speed-of-the-voronoi-projection. See also the whole collection of notebooks in https://observablehq.com/@fil/polyhedral-projections-with-d3-geo-polygon?collection=@fil/projections
